Add support for Xcode 15.0 (Conan 1.x)

janosvitok opened the pull request:
Changelog: Feature: Add support fox Xcode 15.
Docs: Omit

- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.

#14494

janosvitok committed:
Add support for Xcode 15

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=14533) <br/>All committers have signed the CLA.

memsharded reviewed (commented):
Hi @janosvitok 

Thanks for your contribution.

At this moment we are only accepting changes in the Conan 1.X that are strictly necessary. With XCode and all these new versions being beta, this PR seems a bit beyond what is expected.

Adding the minimum necessary OSX (``Macos``) and XCode versions when Apple finally releases them could be ok and necessary due to Apple forced updates, but all the newer versions for all the Apple platforms like iOS, watchOS might not be that critical or necessary.

I suggest putting this on hold until Apple releases mandate the updates, and restrict them to the bare minimum to not break, and target this PR to the ``release/2.0`` branch for the Conan 2.0 releases.

janosvitok commented:
Hi @memsharded,

our situation is this:

- we need to prepare our program so that it works with the new version of iOS once it is released. To compile for iOS beta, we need to use Xcode beta. I don't really mind the various iOS versions, I've added them for completeness sake. 

- we store our package in gitlab conan repository, and unfortunately, gitlab [does not support conan 2.x yet](https://gitlab.com/gitlab-org/gitlab/-/issues/348228). So we have to stay with conan 1.x for now. We can manage this by editing the settings manually, if xcode 15 support does not make it to the 1.x series.

There is approximately [1 month](https://9to5mac.com/2023/08/19/ios-17-new-features-release-date-devices/#h-when-will-ios-17-be-released-to-everyone) left before Apple releases iOS 17. They will release Xcode 15 about the same time, since its crucial to develop apps for iOS 17. It would be great if support for Xcode 15 in conan will be released prior to Xcode 15 release, otherwise people will have to manually edit their conan settings files.

What do you think? Should I remove iOS/watchOS versions? Should I split it to two PRs: one for 1.x with Xcode 15 only, and one for 2.x with all the versions?

memsharded commented:
Thanks very much for the feedback @janosvitok!

Let me explain our point of view 🙂 : Conan 2.0 beta versions are already more than 1 year old now (https://docs.conan.io/2/changelog.html) and alphas date back to Nov-2021 (https://github.com/conan-io/conan/releases/tag/2.0.0-alpha1). I understand the will and need to use latest versions of other tools, but we cannot keep indefinitely adding things to Conan 1.X to support them, as it is equally important to update to Conan 2.0, also because ConanCenter will keep moving forward and stop building some 1.X configurations to build them for 2.0, as the resources are not infinite, and eventually 1.X support will be frozen in ConanCenter. 

Also every review takes time, and then things need to be merged forward to 2.0 (there are always conflicts, etc). Then things might start failing with the latest iOS version with the latest CMake version, because some policy changed, and users will keep asking for 1.X support, because the settings are already defined built-in. At some points limits have to be defined, the effort of developing, migrating and maintaining Conan 1.X and 2.0 simultaneously is huge and exhausting, and can't be sustained forever.

> we store our package in gitlab conan repository

The API v2 has been out for more than 4 years, and it is all is needed to run Conan 2.0. Conan 2.0 do not need a new API or new support, just the revisions-capable API v2 that was introduced more than 4 years ago. So this is unfortunate, but this is something that should be fixed from their side, I suggest pushing them until they do. And if they don't fix it, we have the completely free ArtifactoryCE, it is not that challenging to run and maintain in your own server, I'd suggest giving it a try.

In summary, we all need to move forward, and the only sustainable way is to stop supporting newer platforms in older versions of the tools. As users wanting to stay in Conan 1.X can easily add their own ``settings.yml`` and share it with ``conan config install`` (which always has been the recommended way to share profiles and other stuff), this shouldn't be a blocker. So I'd suggest the following path:

- Do a PR with the full thing to Conan 2.0
- Do a PR with the bare minimum to not break when XCode 15 is out: I think at this stage the only really breaking thing would be the ``apple-clang`` 15.0 version, as all the other things are opt-in
- Users wanting to use newer versions of iOS, Macos sdk, etc, with Conan 1.X will need to add them to their own ``settings.yml``

Please let me know what you think.

janosvitok commented:
Sounds reasonable. 

I'm aware that it's gitlab fault, but I'm afraid we're too small to force them to do something (unless we implement it ourselves, which we don't have capacity for). Edit: It seems that there is a work in progress: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/121860

So I will:

 - strip everything but Xcode 15 from this PR
 - create another PR targeted at release/2.0 with all changes

janosvitok commented:
Note: Xcode 15 was added to 2.x branch in #14302.

H-G-Hristov commented:
Thanks. We'll also have to live with Conan 1.x for a long while due to a large amount of customized recipes.
So for us getting new Xcode version and C++ language versions (C++26) in Conan 1.x is essential.

H-G-Hristov commented on conans/test/unittests/client/build/cpp_std_flags_test.py:
```diff
@@ -217,6 +217,13 @@ def test_apple_clang_cppstd_flags(self):
+        self.assertEqual(_make_cppstd_flag("apple-clang", "15.0", "gnu17"), "-std=gnu++17")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "15.0", "20"), "-std=c++20")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "15.0", "gnu20"), "-std=gnu++20")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "15.0", "23"), "-std=c++2b")
+        self.assertEqual(_make_cppstd_flag("apple-clang", "15.0", "gnu23"), "-std=gnu++2b")
```
Doesn't AppleClang support `-std=c++23` already?

janosvitok replied:
```bash
$ echo '' | ./clang -x c++ --std=c++23 -E -
error: invalid value 'c++23' in '--std=c++23'
note: use 'c++98' or 'c++03' for 'ISO C++ 1998 with amendments' standard
note: use 'gnu++98' or 'gnu++03' for 'ISO C++ 1998 with amendments and GNU extensions' standard
note: use 'c++11' for 'ISO C++ 2011 with amendments' standard
note: use 'gnu++11' for 'ISO C++ 2011 with amendments and GNU extensions' standard
note: use 'c++14' for 'ISO C++ 2014 with amendments' standard
note: use 'gnu++14' for 'ISO C++ 2014 with amendments and GNU extensions' standard
note: use 'c++17' for 'ISO C++ 2017 with amendments' standard
note: use 'gnu++17' for 'ISO C++ 2017 with amendments and GNU extensions' standard
note: use 'c++20' for 'ISO C++ 2020 DIS' standard
note: use 'gnu++20' for 'ISO C++ 2020 DIS with GNU extensions' standard
note: use 'c++2b' for 'Working draft for ISO C++ 2023 DIS' standard
note: use 'gnu++2b' for 'Working draft for ISO C++ 2023 DIS with GNU extensions' standard
$ ./clang --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: x86_64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Applications/Xcode_15_beta_6.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/.
```

See also https://github.com/conan-io/conan/issues/14494#issuecomment-1681625657

H-G-Hristov replied:
> ```shell
> ```shell
> $ echo '' | ./clang -x c++ --std=c++23 -E -
> error: invalid value 'c++23' in '--std=c++23'
> note: use 'c++98' or 'c++03' for 'ISO C++ 1998 with amendments' standard
> note: use 'gnu++98' or 'gnu++03' for 'ISO C++ 1998 with amendments and GNU extensions' standard
> note: use 'c++11' for 'ISO C++ 2011 with amendments' standard
> note: use 'gnu++11' for 'ISO C++ 2011 with amendments and GNU extensions' standard
> note: use 'c++14' for 'ISO C++ 2014 with amendments' standard
> note: use 'gnu++14' for 'ISO C++ 2014 with amendments and GNU extensions' standard
> note: use 'c++17' for 'ISO C++ 2017 with amendments' standard
> note: use 'gnu++17' for 'ISO C++ 2017 with amendments and GNU extensions' standard
> note: use 'c++20' for 'ISO C++ 2020 DIS' standard
> note: use 'gnu++20' for 'ISO C++ 2020 DIS with GNU extensions' standard
> note: use 'c++2b' for 'Working draft for ISO C++ 2023 DIS' standard
> note: use 'gnu++2b' for 'Working draft for ISO C++ 2023 DIS with GNU extensions' standard
> $ ./clang --version
> Apple clang version 15.0.0 (clang-1500.0.40.1)
> Target: x86_64-apple-darwin22.6.0
> Thread model: posix
> InstalledDir: /Applications/Xcode_15_beta_6.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/.
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> See also #14494 (comment)
> ```

Thank you. It looks like this was resolved only in Clang 17:

> error: invalid value 'c++27' in '-std=c++27'
note: use 'c++98' or 'c++03' for 'ISO C++ 1998 with amendments' standard
note: use 'gnu++98' or 'gnu++03' for 'ISO C++ 1998 with amendments and GNU extensions' standard
note: use 'c++11' for 'ISO C++ 2011 with amendments' standard
note: use 'gnu++11' for 'ISO C++ 2011 with amendments and GNU extensions' standard
note: use 'c++14' for 'ISO C++ 2014 with amendments' standard
note: use 'gnu++14' for 'ISO C++ 2014 with amendments and GNU extensions' standard
note: use 'c++17' for 'ISO C++ 2017 with amendments' standard
note: use 'gnu++17' for 'ISO C++ 2017 with amendments and GNU extensions' standard
note: use 'c++20' for 'ISO C++ 2020 DIS' standard
note: use 'gnu++20' for 'ISO C++ 2020 DIS with GNU extensions' standard
note: use 'c++23' for 'ISO C++ 2023 DIS' standard
note: use 'gnu++23' for 'ISO C++ 2023 DIS with GNU extensions' standard
note: use 'c++2c' or 'c++26' for 'Working draft for C++2c' standard
note: use 'gnu++2c' or 'gnu++26' for 'Working draft for C++2c with GNU extensions' standard

janosvitok replied:
I've checked beta 7 and the result is the same (more or less expected).
